### PR TITLE
xdsl: provide custom __str__ and __format__ functions for attrs and ops

### PR DESCRIPTION
**Pull request by superlopuh**
This test shows the new output, I think this would be useful for debugging, and not much information lost since the printed output is supposed to be unambiguous anyway.

``` python
def test_descriptions():
    a = Constant.from_int_and_width(1, 32)

    assert str(a.value) == '1 : !i32'
    assert f'{a.value}' == 'IntegerAttr(1 : !i32)'

    assert str(a) == '%0 : !i32 = arith.constant() ["value" = 1 : !i32]'
    assert f'{a}' == 'Constant(%0 : !i32 = arith.constant() ["value" = 1 : !i32])'

    m = ModuleOp.from_region_or_ops([a])

    assert str(m) == '''\
builtin.module() {
  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
}'''

    assert f'{m}' == '''\
ModuleOp(
\tbuiltin.module() {
\t  %0 : !i32 = arith.constant() ["value" = 1 : !i32]
\t}
)'''

```

Would there be any downsides to printing like this?

### CODE REVIEW

File: tests/test_ir.py [hunk 1]
```diff
@@ -104,9 +105,13 @@ def test_op_clone():
     a = Constant.from_int_and_width(1, 32)
     b = a.clone()
 
+    b_value = b.value
```
Comment (georgebisbas) [inline tests/test_ir.py:108]: This line adds extra confusion I think

**Review by georgebisbas — COMMENTED** [review]
Is this PR in the direction of nearly substituting printers?

**Comment by math-fehr** [comment]
This looks fine on these examples, but I'm a bit scared of what happens when:
* We have large programs. Do you want to print 1k+ lines of code when there is a bug? Or should we print something like : "1000 operations in region". Note that we might have a different answer for `str` and `repr`.
* We have operands that are not printed yet. Should we print `%<<UNKOWN>>`, or should we get an error? Same here, we might want something different for `str` and `repr`.

My feeling is that we should add the `%<<UNKNOWN>>` to `__str__` and `__repr__`, and we should not print the entire regions in `__repr__`, because that will make them too unreadable.

What do you think?

**Comment by AntonLydike** [comment]
I think this is a super valuable PR, and this feature would improve the development experience. I think we should think about what @math-fehr said, but this is a great first step. Imho this doesn't need to be perfect right from the start, and collapsing of large regions can always be done later.

I'm very happy to see progress on this front!

**Review by AntonLydike — APPROVED** [review]

**Comment by martin-luecke** [comment]
> This looks fine on these examples, but I'm a bit scared of what happens when:
> 
> * We have large programs. Do you want to print 1k+ lines of code when there is a bug? Or should we print something like : "1000 operations in region". Note that we might have a different answer for `str` and `repr`.
> 
> * We have operands that are not printed yet. Should we print `%<<UNKOWN>>`, or should we get an error? Same here, we might want something different for `str` and `repr`.
> 
> 
> 
> My feeling is that we should add the `%<<UNKNOWN>>` to `__str__` and `__repr__`, and we should not print the entire regions in `__repr__`, because that will make them too unreadable.
> 
> 
> 
> What do you think?

Really cool, to see progress on this. For prototyping I remember regularly changing __str__ abd __repr__ to my needs.

I think the both behaviors are very valuable and I think there were different points in time where I needed one of them.

Does it make sense to have e.g a "debug" flag that enables the verbose printing?

**Comment by superlopuh** [comment]
@math-fehr, These are valid concerns, but I'm not sure how bad it would be in practice. My understanding is that if a user of a compiler written in xDSL sees a direct printout of an op or attribute then there's a bug somewhere. Almost all our errors are more carefully constructed. For debugging, having the full printout-even if it takes a 100 terminal screens, is still good information.

**Comment by superlopuh** [comment]
Oh and that means that a debug flag is also sort of implied, the only time to print an op directly is when debugging

**Comment by superlopuh** [comment]
I initially wanted to implement `__str__` and `__repr__`, but went with `__format__` instead of repr, for the print function working better. The main reason is the docstring for `__repr__` saying that the following expression usually evaluates to `True` for Python builtins: `eval(repr(x)) == x`. It would be cool for xDSL to stick to this, and can be done in a separate PR

File: xdsl/ir.py [hunk 2]
```diff
@@ -630,6 +630,29 @@ def __eq__(self, other: object) -> bool:
...
+        res = StringIO()
+        printer = Printer(stream=res)
+        printer.print_op(self)
+        desc = res.getvalue()
+        # Printer currently prints a new line when printing an operation.
+        # We don't want this behaviour when printing ops inline, so remove
+        # the trailing newline for now
+        last_char_is_newline = not desc or desc[-1] == "\n"
+        assert last_char_is_newline, "If the format changes, update this function"
+        desc = desc[:-1]
```
Comment (webmiche) [inline xdsl/ir.py:644]: I would very much appreciate if we could fix this in a later PR. Having this kind of code in the repo feels bad to me.

**Review by webmiche — APPROVED** [review]
LGTM

**Comment by martin-luecke** [comment]
> Oh and that means that a debug flag is also sort of implied, the only time to print an op directly is when debugging

I mean to say that I think a flag to switch the behavior would be cool. Maybe "-vebose-logging", "-debug-verbose" or something

**Comment by georgebisbas** [comment]
> Does it make sense to have e.g a "debug" flag that enables the verbose printing?

I like this idea.

In devito we use something starting from here: https://github.com/devitocodes/devito/blob/master/devito/logger.py

**Comment by georgebisbas** [comment]
https://docs.python.org/3/library/logging.html

File: xdsl/ir.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/ir.py:644]: #450

**Comment by superlopuh** [comment]
Oh I agree wrt logging in general. My understanding is that it's orthogonal to these changes. The description and logging are somewhat different things.

**Review by math-fehr — APPROVED** [review]
I agree, this doesn't need to be perfect right away, let's fix some issues in following PRs.
This is way better than what we currently have ;)

**Pull request merged**